design/design-doc-custom-composition.md

```diff
@@ -0,0 +1,179 @@
+                description: Crossplane AWS provider credentials to use.
+                properties:
+                  name:
+                    type: string
+                required:
+                - name
+            required:
+            - region
+            - providerConfigRef
+```
+
+User can have multiple `Composition` or custom compositions that target this new
+type. Let's assume they would like to write a cdk8s app and use it as the composition
+engine. They'd need to write the cdk8s app that will take the following resources
+as input:
+  * `PrivateNetwork` custom resource.
+  * `CustomComposition` custom resource.
+  It will return all the resulting resources as output.
+
+This conversion function will be wrapped by an HTTP server and pushed as an image
+whose `ENTRYPOINT` is to start that HTTP server. The input will come as POST
+request and the body of the result will consist of all the output resources.
+
+At this point, we have our cdk8s app that will act as composition engine. Now,
+let's go through how the consumers of that app will be able to use it.
+
+### Consumption Experience
+
+Now that our composition engine is ready, we'll instruct Crossplane to install
+that composition engine and send the appropriate requests. In order to do that,
+the following `CustomComposition` will be created:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: CustomComposition
+metadata:
+  name: acmenetwork.acme.org
+spec:
+  writeConnectionSecretsToNamespace: crossplane-system
+  reclaimPolicy: Delete
+  compositeTypeRef:
+    apiVersion: aws.example.org/v1alpha1
+    kind: PrivateNetwork
+  image: acme/network-cdk8s:1.0.0
+```
+
+Once this `CustomComposition` is created, a new `Deployment` will be created by
+Crossplane with the following properties:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: acmenetwork.acme.org
+  labels:
+    app: acmenetwork.acme.org
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: acmenetwork.acme.org
+  template:
+    metadata:
+      labels:
+        app: acmenetwork.acme.org
+    spec:
+      containers:
+      - name: custom-composition
+        image: crossplane/custom-composition:0.0.1
+        env:
+        # COMPOSITE_TYPE will let the custom-composition controller know what type
+        # it should watch.
+        - name: COMPOSITE_TYPE
+          value: privatenetworks.aws.example.org
+        # CUSTOMCOMPOSITION_NAME is the name of the CustomComposition that will
+        # be sent over to composition engine alongside composite instance. 
+        - name: CUSTOMCOMPOSITION_NAME
+          value: acmenetwork.acme.org
+      - name: composition-engine
+        image: acme/network-cdk8s:1.0.0
+        ports:
+        - containerPort: 80
+```
+
+After this `Deployment` is up, the container named `custom-composition` will start
+watching `PrivateNetwork` types and it will reconcile only the ones with the following
+reference:
```

Comment:
This seems to imply that the custom composition controller would be separate from core Crossplane. Why is this? Would the custom composition controller change or remain the same if the composition engine changed (e.g. from cdk8s to kustomize)?

Reply:
It'd remain the same. Having both of them in the same `Pod` saves us from in-cluster networking, which could require TLS handling and expand the attack surface, thought that's manageable. But the key point is what entity will create the output YAMLs. If it's core Crossplane pod, then that means cdk8s app can do anything it wants in the cluster since Crossplane has `cluster-admin` and I'd feel really uncomfortable doing that because then `CustomComposition` permission would be equal to root, which is not immediately clear to the user. So, having the deployer together with cdk8s app could give the ability to user to limit what cdk8s app can ever deploy via RBAC specific to its `ServiceAccount`.